### PR TITLE
[Relax][Op] Add repeat, tile, conv2d_transpose

### PR DESCRIPTION
**Pull request by Ubospica**
This PR adds these operators:
- `relax.repeat`
- `relax.tile`
- `relax.conv2d_transpose`
	- This operator is intended to find the gradient of conv2d w.r.t its input. For details see [pytorch document](https://pytorch.org/docs/stable/generated/torch.nn.ConvTranspose2d.html), and the document in `python/tvm/relax/op/nn/nn.py`.
	- Now TOPI support of conv2d_transpose is limited. It does not support dilations; layouts other than default; symbolic `output_padding`.

These operators are serving for #130

### CODE REVIEW

File: tests/python/relax/test_transform_legalize_ops_manipulate.py [hunk 1]
```diff
@@ -888,5 +888,136 @@ def main(x: R.Tensor(("a", "b", "c"), "float32")) -> R.Tensor(("b", 1), "float32
     tvm.ir.assert_structural_equal(mod, Expected)
 
 
+def test_repeat():
+    # fmt: off
```
Comment (SiriusNEO) [inline tests/python/relax/test_transform_legalize_ops_manipulate.py:892]: Do we need `test_repeat_symbolic`?

File: src/relax/op/tensor/manipulate.h [hunk 2]
```diff
@@ -133,6 +133,33 @@ Expr collapse_sum_like(Expr data, Expr collapse_target);
...
 
+/*!
+ * \brief Repeats elements of an array.
+ * \param data The input tensor.
+ * \param repeats The number of repetitions.
+ * \param axis The axis along which to repeat values. The negative numbers are interpreted counting
+ * from the backward. By default, use the flattened input array, and return a flat output array.
+ * \return The computed result.
+ */
+Expr repeat(Expr data, PrimExpr repeats, Optional<Integer> axis = NullOpt);
```
Comment (SiriusNEO) [inline src/relax/op/tensor/manipulate.h:144]: Seems there is no need to set the default value `= NullOpt`

File: tests/python/relax/test_transform_legalize_ops_manipulate.py [hunk 3]
```diff
@@ -18,7 +18,7 @@
 import pytest
 import tvm
 from tvm.relax.transform import LegalizeOps
-from tvm.script import relax as R, tir as T
+from tvm.script import relax as R, tir as T, ir as I
```
Comment (SiriusNEO) [inline tests/python/relax/test_transform_legalize_ops_manipulate.py:21]: Not sure, but I found in other tests we use `@tvm.script.ir_module`

Comment (Ubospica) [reply tests/python/relax/test_transform_legalize_ops_manipulate.py:21]: The problem is that the printer prints `I.ir_module` by default 

Confirmed with Ruihang that they can substitude each other :)

File: src/relax/op/tensor/manipulate.h (hunk 2 above)
Comment (Ubospica) [reply src/relax/op/tensor/manipulate.h:144]: If we use these operators in C++, would providing a default value make it more convenient?

Comment (SiriusNEO) [reply src/relax/op/tensor/manipulate.h:144]: In this case maybe we can add a default value for `repeat`. And also add for other ops with such Optional argument.

File: tests/python/relax/test_transform_legalize_ops_manipulate.py [hunk 4]
```diff
@@ -888,5 +888,214 @@ def main(x: R.Tensor(("a", "b", "c"), "float32")) -> R.Tensor(("b", 1), "float32
     tvm.ir.assert_structural_equal(mod, Expected)
 
 
+def test_repeat():
+    # fmt: off
+    @I.ir_module
+    class Repeat:
+        @R.function
+        def main(x: R.Tensor((3, 2, 3), "float32")):
```
Comment (Ubospica) [inline tests/python/relax/test_transform_legalize_ops_manipulate.py:896]: I ignored some annotations in the Before module. The problem is, if we specify type annotations, `InferStructInfo` checking will be skipped. Then some errors in hand-written struct_infos may not be found.

For example, `test_nll_loss` in `test_transform_legalize_ops_nn.py`

Comment (Ubospica) [reply tests/python/relax/test_transform_legalize_ops_manipulate.py:896]: What about adding this feature:

Always check struct_infos for operators, tuples, tuplegetitems, etc. when normalizing, even though it is specified?

It seems not hard to implement for us

@MasterJH5574

**Comment by MasterJH5574** [comment]
> * `0` parameter for `reshape`.
>   
>   * If there exist 0-s in the shape, that denotes the dimension of the output shape equals to the corresponding input dimension.
>   * The dimension is counted from left to right.
>   * This a feature implemented in `relay.reshape`, paddlepaddle, matlab, etc.

@Ubospica I don’t think this is a standard behavior. None of the mainstream frameworks and specifications (NumPy, PyTorch, TF, DataAPI) has this specification. We should align our API with them. If, at some time point, you need a frontend importer say from paddlepaddle, you can handle the zeros in the importer.

Update: No need to align with Relay. It is the mainstream frameworks and specifications that we should get closer to.

**Comment by MasterJH5574** [comment]
> Adds these modifications to Relax operators and parser:

Could you do the fixes this as separate PRs? Small patches, with one for each fix with corresponding, are preferable and more manageable than a huge PR. Try also send some to tlc or unity if the site is already there.

**Comment by tqchen** [comment]
closing for now in favor of mlc branch
